[FIX JENKINS-32485] New extension point for contributing usage statistics / preparatory for JENKINS-26466

batmat opened the pull request:
This pull request introduces [a new feature/JENKINS-32485](https://issues.jenkins-ci.org/browse/JENKINS-32485): now the UsageStatistics class doesn't know the stats it's going to send anymore.

It delegates to a new UsageStatisticsContributor extension point, and iterates through the implementations to find what to send.

---

This PR also includes the refactoring for the coupling between UsageStatistics and ArchitectureMonitor (preparatory for [JENKINS-26466](https://issues.jenkins-ci.org/browse/JENKINS-26466)). 

I hesitated to include it in another PR, but finally thought it would be good to:
- illustrate what it would be useful for directly here (i.e. breaking some undesirable relationships between some classes in this case). 
- And also proving that it actually enables it through a concrete usage and I haven't missed something to make it usable.

Thanks for your feedback. Thanks to @jglick for his help on the design.

Ping @jenkinsci/code-reviewers

batmat committed:
[FIX JENKINS-32485] Extension Point for contributing usage statistics

Existing usage statistics has for the moment simply been moved to a new
@Extension class.

batmat committed:
[JENKINS-26466] Decouple UsageStatistics from ArchitectureMonitor

By introducing a bridge class that will contribute stats coming from the
node_monitors.Architecture class to the usage statistics subsystem.

daniel-beck commented on core/src/main/java/hudson/model/UsageStatistics.java:
```diff
@@ -126,51 +125,19 @@ private RSAPublicKey getKey() {
-                if(item.getDescriptor()==d)
-                    cnt++;
+        for (UsageStatisticsContributor contributor : UsageStatisticsContributor.all()) {
+            JSONObject contributed = contributor.getStatData();
```
Wrap in try/catch for robustness.

batmat replied:
@daniel-beck OK, on it now. 

Question about robustness: do you think the code should just log an error if a `UsageStatisticsContributor` tries to override a key instead of throwing an IllegalStateException as it does now? 

IMO, to be consistent with catching exception here, it would seem more logical to try and continue instead in general?

batmat replied:
Done both: 
* catch(Exception) now around the .getStatDatat() call for each _contributor_
* if a _contributor_ tries to override a key, just ignore it and log a severe message

daniel-beck commented on core/src/main/java/hudson/model/UsageStatisticsContributor.java:
```diff
@@ -0,0 +1,59 @@
+
+/**
+ * Extension point to contribute usage statistics to the {@link UsageStatistics}
+ * class and associated infrastructure.
```
Should document requirement for corresponding change to https://github.com/jenkinsci/infra-statistics

batmat replied:
Can you please elaborate a bit more @daniel-beck ? Do you mean we should write some form of (light) specification for the required/optional fields to clarify the relationship and expectations between here and infra-statistics? Thanks

daniel-beck replied:
@batmat Implementations of this extension point should have a corresponding change in infra-statistics to process the data gathered. Or do you just want to have plugins dump data and then forget about it?

batmat replied:
Understood now I think. I thought you wanted to make a somehow generic spec as to how handle that kind of data on the infra-statistics project. 

I'll add a comment about that in that Javadoc (and also the privacy heavy warning discussed on the JIRA). The fact it's **strongly recommended** to have the community at large, and the security officer in particular review a new implementation (?).

batmat replied:
Done.

daniel-beck commented on core/src/main/java/hudson/model/UsageStatisticsContributorForCore.java:
```diff
@@ -0,0 +1,71 @@
+import net.sf.json.JSONObject;
+
+@Extension
+public class UsageStatisticsContributorForCore extends UsageStatisticsContributor {
```
Weird class name. Maybe `CoreUsageStatistics`?

batmat replied:
Yeah. Renamed. 
Originally named it to have it visually alongside UsageStatistics while starting the modifications. I admit it was not great.

daniel-beck commented on core/src/main/java/hudson/node_monitors/NodesDetailsUsageStatisticsContributor.java:
```diff
@@ -0,0 +1,64 @@
+ * 
+ * @see ArchitectureMonitor
+ */
+public class NodesDetailsUsageStatisticsContributor extends UsageStatisticsContributor {
```
`@Extension`

batmat replied:
Fixed. Thanks

olivergondza commented:
Is this supposed to change the data layout sent by core by default? If not, it should have a comparative test to make sure, the new json is semantically equivalent to the old one.

batmat commented:
@olivergondza Not sure what you have in mind. I actually added tests to UsageStatisticsTest for that purpose as first commit, do you think it should still be enriched in this regard? 
Or do you think of something totally differerent?

It's indeed expected to be structurally equivalent, but with different ordering: the 'nodes' attribute is going to be at the end or beginning of the json depending on the extension detection order. 

---
Side note: is there any documentation about the rules of the extension detection order? My understanding is that there's no order guaranty apart from the one provided by Extension.ordinal, right? 
Side side note: are Extensions declared in core going to be generally detected first, or not at all? (only depending on classpath ordering I suppose).

olivergondza commented:
Comparing the output on text level, only values of `installed` and `version` has changed. Here is my +1.

This is what I suggested for the unittest to do (so if `pugins` list would be always empty after the refactoring for instance, the tests would detect that).

batmat commented:
OK @olivergondza I'll add more harnessing to that. 

Like I said in the test comment, this will couple the test to the `UsageStatistics` internals even more, but that's also true that it's certainly preferable to watch it very carefully as statistics generation depends very much on the format of all that.

batmat commented:
Just pushed an updated version:
@daniel-beck I think I addressed all your remarks, is there anything else you still think is missing?

@olivergondza added a full check of the `plugins`, `jobs` and `nodes` json attributes. As stated in the comments, maybe it's tying a bit too strong the test with that part of the core, but as this format should really not change to not break _infra-statistics_ I think it's indeed worth it.

batmat commented:
Hi!
I'd like to make [JENKINS-26466](https://issues.jenkins-ci.org/browse/JENKINS-26466) go forward. 

So if there's still anything missing, please just tell me I'll work on fixing it.

ping @jenkinsci/code-reviewers for more eyes

Thanks!

jglick commented:
> there's no order guaranty apart from the one provided by Extension.ordinal, right?

IIRC extensions with the same ordinal are sorted lexicographically, for deterministic behavior, but you can consider this order arbitrary.

> are Extensions declared in core going to be generally detected first

No, location of definition should be irrelevant.

jglick commented on core/src/main/java/hudson/model/CoreUsageStatistics.java:
```diff
@@ -0,0 +1,71 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package hudson.model;
```
Newly introduced types are generally placed in `jenkins.*`. I do not feel strongly about it, just FYI.

jglick commented on core/src/main/java/hudson/model/UsageStatistics.java:
```diff
@@ -126,51 +125,38 @@ private RSAPublicKey getKey() {
+            try { // robustness: a sole contributor shouldn't be able to interrupt the whole stats retrieval process
+                contributed = contributor.getStatData();
+            } catch (Exception e) {
+                LOGGER.warning(String.format("'%s' threw a '%s 'exception: '%s'.", contributorClass, e, e.getMessage()));
```
Better to say

```java
} catch (RuntimeException e) {
    LOGGER.log(Level.WARNING, null, e);
}
```

The information about the class is almost surely redundant given a stack trace.

jglick commented on core/src/main/java/hudson/model/UsageStatisticsContributor.java:
```diff
@@ -0,0 +1,79 @@
+     * @return all the {@link UsageStatisticsContributor} instances.
+     */
+    public static ExtensionList<UsageStatisticsContributor> all() {
+        return Jenkins.getActiveInstance().getExtensionList(UsageStatisticsContributor.class);
```
You can use `ExtensionList.lookup` more concisely.

jglick commented on core/src/main/java/hudson/model/UsageStatisticsContributor.java:
```diff
@@ -0,0 +1,79 @@
+     * Gets the information to be contributed to the usage statistics subsystem.
+     *
+     * <p>
+     * <strong>Beware that using an existing key will throw an exception. Because we don't want a plugin to be able to
```
Well, will not technically “throw an exception”. Will be logged.

jglick commented on core/src/main/java/hudson/node_monitors/NodesDetailsUsageStatisticsContributor.java:
```diff
@@ -0,0 +1,66 @@
+            n.put("os", descriptor.get(c));
+            nodes.add(n);
+        }
+        o.put("nodes", nodes);
```
This means that unless you have node monitors installed, you will get no node information at all. Is that really what you want?

An alternative would be to change the extension point definition, so that the caller first constructs an empty `JSONObject`, then asks extensions to insert data into it. Thus you would not be limited to adding top-level keys. Obviously plugins need to coöperate in this case and be written defensively, but any implementation of this extension point needs to be written with care anyway to avoid sending private data.

batmat replied:
Right. IIUC, you mean that basically only L59&L60 should really be _there_, and the rest be left in the `CoreUsageStatistics` class?

Since only those lines
```
DescriptorImpl descriptor = j.getDescriptorByType(DescriptorImpl.class);
n.put("os", descriptor.get(c));
```
are indeed coupling UsageStatistics to `node_monitors`. 

Totally makes sense, I'll see how I can rework that in the next few days. Cause indeed this wouldn't be an issue in short term as we plan to bundle the extracted plugin anyway, but would be a bigger risk to lower the richness of retrieved stats in the future once that plugin may not be bundled anymore.

batmat commented:
Just addressed all @jglick comments in the last commits.

(Considered at first amending the existing commits, but then I thought that might be easier to review and to possibly discuss if I do many commits. I'll squash those into more concise (like the initial 4) when review is done.)

batmat commented:
Just fixed the merge conflict after 2.0 got merged to `master`.

batmat commented:
Ping @jglick as one of the advisors of the design.
Ping @jenkinsci/code-reviewers to help it go forward. 

Reminder: the goal of that PR is to make a step to be able to extract node_monitors into a dedicated plugin ([JENKINS-26466](https://issues.jenkins-ci.org/browse/JENKINS-26466)).

Thanks.

daniel-beck commented on core/src/main/java/hudson/model/UsageStatistics.java:
```diff
@@ -123,58 +127,42 @@ private RSAPublicKey getKey() {
+            Class<? extends UsageStatisticsContributor> contributorClass = contributor.getClass();
+            LOGGER.finest(String.format("Calling '%s' usage stats contributor",  contributorClass));
+
+            try { // robustness: a sole contributor shouldn't be able to interrupt the whole stats retrieval process
```
Robustness would catch `Error` (possibly class loading related?)

batmat replied:
Catching `Throwable` generally seems like a giant code smell to me. 
Not sure it would be justified here :-/, it would catch `OutOfMemoryError`s and many important issues. So yeah, this is not robust in the sense it's currently gonna catch everything, but everything that seems reasonable to catch.
So, I'm not really sure we should do this.

Looking at the code, I do see indeed many instances of that pattern, so maybe it's just me not used to it, and it's just normal for Jenkins core code as it's a container that should aggressively protect instances from others. I'll think about it.

Any opinion @jenkinsci/code-reviewers ?

daniel-beck commented on core/src/main/java/jenkins/model/statistics/UsageStatisticsContributor.java:
```diff
@@ -0,0 +1,82 @@
+ * processed ;</li>
+ * <li><strong>WARNING:</strong> Because there can be security/privacy implications, before adding a new implementation
+ * of this extension point in your plugin, it is <strong>strongly recommended</strong> to get your code reviewed by the
+ * community in general, and by the
```
Maybe a useful process would be to only accept core changes once the `infra-statistics` change is in, and _that_ is the closely reviewed one?

daniel-beck replied:
I'm not sure this should be a security officer job here. Users wishing to add to this probably don't care.

daniel-beck commented:
Would be great if this accounted for dynamically building up the documentation on usage statistics, so the help would show which plugin (automatic) collects which data and for which purpose (manual).

daniel-beck commented:
@oleg-nenashev Why did you assign me?

rtyler commented:
I would like this to be put on hold until some more discussion happens on [this infra list thread](http://lists.jenkins-ci.org/pipermail/jenkins-infra/2016-June/000763.html)

batmat commented:
Closing as per http://meetings.jenkins-ci.org/jenkins-meeting/2016/jenkins-meeting.2016-06-22-18.00.log.html#l-170
